### PR TITLE
Add Display Recording and Audio Dumping to Desktop

### PR DESCRIPTION
**Pull request by RisingFog**
This is a work in progress port of Dolphin's frame dumping utility for PPSSPP.

Here is a current TODO list of what is left to be done:
- [x] Fix incorrect video output
- [x] Create menu options (Win32 at least)
- [x] Add OSD messages (not currently required)
- [x] Fix minor issues with video dumping (crashing on resolution changes)
- [x] Add audio dumping

~~I will be writing audio dumping separate to this PR.~~

### CODE REVIEW

**Comment by xnamkcor** [comment]
Is this based on Dolphin's 1 frame dump, or did Dolphin get a sequential frame dump?

**Comment by RisingFog** [comment]
@xnamkcor I added splitting video files to Dolphin post 5.0, so it's essentially a port of what is currently had in Dolphin.

**Comment by xnamkcor** [comment]
@RisingFrog. I downloaded the latest automatic build as of my comment 3 hours ago and I saw no video recording options. Is this a feature that is not yet added to the builds to be automatically built? Was video recording already an option in the automatic builds? Or did you add the resolution change handling before the video recording was added?

**Comment by RisingFog** [comment]
@xnamkcor for PPSSPP or Dolphin? For Dolphin, you have to enable Frame and Audio dumps from the Movie menu, then it will dump the files to either a default location, or a location of your choice set in the settings. 

PPSSPP only has the frame dump option in the movie menu with the option for lossless (FFV1). It's not officially built into PPSSPP yet, you'll need to compile it yourself until it's merged.

**Comment by xnamkcor** [comment]
@RisingFog Ok. I'll wait until it's merged. I thought I recalled something weird about Dolphin's frame dump. After looking, it seems like it asks which codec to use and "Full Frames(Uncompressed)" is an option now. I may not have checked since before 5.x. If the same is planned for PPSSPP, I'm happy.

File: Core/System.cpp [hunk 1]
```diff
@@ -593,6 +593,8 @@ std::string GetSysDirectory(PSPDirectories directoryType) {
 			return g_Config.appCacheDirectory;
 		}
 		return g_Config.memStickDirectory + "PSP/SYSTEM/CACHE/";
+	case DIRECTORY_VIDEO_DUMP:
+		return g_Config.memStickDirectory + "PSP/FRAMEDUMP/";
```
Comment (unknownbrackets) [inline Core/System.cpp]: I guess `PSP/VIDEO/` is probably a more PSP-ish place to put this?

-[Unknown]

File: Core/SaveState.cpp [hunk 2]
```diff
@@ -25,6 +25,7 @@
 #include "Common/FileUtil.h"
 #include "Common/ChunkFile.h"
 
+#include "Core/AVIDump.h"
```
Comment (unknownbrackets) [inline Core/SaveState.cpp]: Is this change still necessary?

-[Unknown]

File: Core/HW/StereoResampler.h [hunk 3]
```diff
@@ -25,6 +25,7 @@
 
 #include "Common/ChunkFile.h"
 #include "Common/CommonTypes.h"
+#include "Core/WaveFile.h"
```
Comment (unknownbrackets) [inline Core/HW/StereoResampler.h]: Can we move this include to the files that need it?  We try to keep includes down to reduce churn and compile times.

-[Unknown]

File: Core/HLE/__sceAudio.h [hunk 4]
```diff
@@ -18,6 +18,7 @@
 #pragma once
 
 #include "sceAudio.h"
+#include "Core/WaveFile.h"
```
Comment (unknownbrackets) [inline Core/HLE/__sceAudio.h]: Can we move this is `__sceAudio.cpp`?

-[Unknown]

File: Core/HLE/__sceAudio.cpp [hunk 5]
```diff
@@ -364,6 +380,15 @@ void __AudioUpdate() {
 
 	if (g_Config.bEnableSound) {
 		resampler.PushSamples(mixBuffer, hwBlockSize);
+		if (m_logAudio)
+		{
+			s16 *clamped_data = new s16[hwBlockSize * 2];
```
Comment (unknownbrackets) [inline Core/HLE/__sceAudio.cpp]: It looks like this memory is never freed.  Should we use a `vector` instead or maybe a member/static global so we don't need to reallocate every frame?

-[Unknown]

File: Core/HLE/__sceAudio.cpp [hunk 6]
```diff
@@ -67,6 +68,9 @@ static int audioHostIntervalCycles;
 
 static s32 *mixBuffer;
 
+WaveFileWriter g_wave_writer;
+bool m_logAudio;
```
Comment (unknownbrackets) [inline Core/HLE/__sceAudio.cpp]: Isn't this a g_?  Or, this should probably be `static`, right?

-[Unknown]

File: Core/CoreTiming.cpp [hunk 7]
```diff
@@ -240,6 +240,11 @@ u64 GetTicks()
 	return (u64)globalTimer + slicelength - currentMIPS->downcount;
 }
 
+u64 GetTicksPerSecond()
```
Comment (unknownbrackets) [inline Core/CoreTiming.cpp]: I think we prefer using `usToCycles` and such to this... can you remove?

-[Unknown]

File: CMakeLists.txt [hunk 8]
```diff
@@ -1190,6 +1190,12 @@ else()
 		ext/disarm.cpp)
 endif()
 
+if (NOT MOBILE_DEVICE)
+	set(CoreExtra ${CoreExtra}
+		Core/AVIDump.cpp
+		Core/AVIDump.h)
```
Comment (unknownbrackets) [inline CMakeLists.txt]: Note: this will get pulled into Qt automatically, so you might need to put `#ifndef MOBILE_DEVICE`.  See here:

https://github.com/hrydgard/ppsspp/blob/master/Qt/Core.pro

(Qt is used for Symbian, for example.)

-[Unknown]

File: CMakeLists.txt [hunk 9]
```diff
@@ -1212,6 +1218,8 @@ add_library(${CoreLibName} ${CoreLinkType}
 	Core/HDRemaster.cpp
 	Core/HDRemaster.h
 	Core/ThreadEventQueue.h
+	Core/WaveFile.cpp
+	Core/Wavefile.h
```
Comment (unknownbrackets) [inline CMakeLists.txt]: If these are always needed, they'll need adding to android as well:

https://github.com/hrydgard/ppsspp/blob/f4e2ca0359e5d2f343cbe8059394ee7b81fd275e/android/jni/Android.mk

-[Unknown]

File: Core/AVIDump.cpp [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+#include "Common/ColorConv.h"
+
+#include "Core/Config.h"
+#include "Core/AVIDump.h"
+#include "Core/System.h"
+#include "Core/Screenshot.h"
+
+#include "GPU/Common/GPUDebugInterface.h"
+#ifdef _WIN32
+#include "GPU/Directx9/GPU_DX9.h"
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: Seems like this include isn't used.  Might be cleaner to remove it and GLES?

-[Unknown]

File: Core/AVIDump.cpp [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+	{
+		pkt->pts = av_rescale_q(s_stream->codec->coded_frame->pts,
+			s_stream->codec->time_base, s_stream->time_base);
+	}
+	if (s_stream->codec->coded_frame->key_frame)
+		pkt->flags |= AV_PKT_FLAG_KEY;
+	pkt->stream_index = s_stream->index;
+}
+
+static const u8 *ConvertBufferTo888RGB(const GPUDebugBuffer &buf, u8 *&temp, u32 &w, u32 &h) {
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: If this is a straight copy, we should probably move it to somewhere common or expose it, rather than duplicating.

We may optimize this or fix bugs in it in the future.  A prime example: we may end up with a new debug buffer format from Vulkan.  Better to just have it one place.

-[Unknown]

File: Core/AVIDump.cpp [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+		}
+		buffer = temp;
+	}
+
+	return buffer;
+}
+
+void AVIDump::AddFrame()
+{
+	GPUDebugBuffer buf;
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: Recommend placing this as a static in the file.  It'll avoid reallocating the buffer each frame.

-[Unknown]

File: Core/AVIDump.cpp [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+	s_src_frame->height = s_height;
+
+	// Convert image from BGR24 to desired pixel format, and scale to initial
+	// width and height
+	if ((s_sws_context =
+           sws_getCachedContext(s_sws_context, w, h, AV_PIX_FMT_RGB24, s_width, s_height,
+                                s_stream->codec->pix_fmt, SWS_BICUBIC, nullptr, nullptr, nullptr)))
+	{
+		sws_scale(s_sws_context, s_src_frame->data, s_src_frame->linesize, 0, h,
+              s_scaled_frame->data, s_scaled_frame->linesize);
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: Please use tabs to indent.  We've included an editorconfig file that can automatically configure your indent styles per project:

https://github.com/hrydgard/ppsspp/blob/master/.editorconfig
http://editorconfig.org/

-[Unknown]

File: Core/AVIDump.cpp [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+		s_sws_context = nullptr;
+	}
+}
+
+void AVIDump::CheckResolution(int width, int height)
+{
+	// We check here to see if the requested width and height have changed since the last frame which
+	// was dumped, then create a new file accordingly. However, is it possible for the height
+	// (possibly width as well, but no examples known) to have a value of zero. This can occur as the
+	// VI is able to be set to a zero value for height/width to disable output. If this is the case,
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: This seems like a very Dolphiny comment.  In PPSSPP, the size can change at any time - but the displayed size is ALWAYS the top left 480x272.  So it may make more sense to just always crop.

Also, there's a "display" framebuffer and a "render" framebuffer.  I'm not sure which one this code is grabbing... the GE debugger uses the "render" one so that might be what's happening here too.  We can probably add a flag.  Just a warning in case there are issues (could cause black screens in some games.)

-[Unknown]

**Comment by unknownbrackets** [comment]
IMHO, I feel like a top level "Movie" option isn't ideal because:

 * Movie might be misunderstood to be related to UMD Video, which we don't support.
 * Movie might be misunderstood to mean in-game videos.
 * Movie probably isn't going to have lots more options?

Maybe "TAS" or "Record" is more appropriate?  I assume that there could be other options related to that which won't necessarily be Movie related, so it might be a better general category for menu options.

-[Unknown]

File: Windows/ppsspp.rc [hunk 15]
```diff
@@ -445,6 +445,13 @@ BEGIN
         MENUITEM "Memory View...",                          ID_DEBUG_MEMORYVIEW
     END
 
+    POPUP "Movie"
+    BEGIN
+        MENUITEM "Dump Frames",                             ID_MOVIE_DUMPFRAMES
+        MENUITEM "Lossless Codec (FFV1)",                   ID_MOVIE_USEFFV1
+		MENUITEM "Dump Audio",                              ID_MOVIE_DUMPAUDIO
```
Comment (unknownbrackets) [inline Windows/ppsspp.rc]: In this file we use spaces.  Sorry, this isn't in the editorconfig yet, should add...

-[Unknown]

File: Windows/ppsspp.rc [hunk 16]
```diff
@@ -445,6 +445,13 @@ BEGIN
         MENUITEM "Memory View...",                          ID_DEBUG_MEMORYVIEW
     END
 
+    POPUP "Movie"
+    BEGIN
+        MENUITEM "Dump Frames",                             ID_MOVIE_DUMPFRAMES
```
Comment (unknownbrackets) [inline Windows/ppsspp.rc]: Hmm, maybe "Record Display" or "Export Screen" would be more clear?  Sorry to nitpick naming.  "Dump frames" makes me think it's going to individually export each frame to a separate file.

-[Unknown]

File: Core/System.cpp (hunk 1 above)
Comment (RisingFog) [inline Core/System.cpp]: Wasn't too sure where to place the dumps, but I guess placing them into VIDEO is fine.

File: Core/SaveState.cpp (hunk 2 above)
Comment (RisingFog) [inline Core/SaveState.cpp]: Nope, will remove it.

File: Core/HW/StereoResampler.h (hunk 3 above)
Comment (RisingFog) [inline Core/HW/StereoResampler.h]: I believe that was leftover from a previous attempt at dumping audio, will move/remove.

File: Core/HLE/__sceAudio.h (hunk 4 above)
Comment (RisingFog) [inline Core/HLE/__sceAudio.h]: Moved.

File: Core/AVIDump.cpp (hunk 13 above)
Comment (RisingFog) [inline Core/AVIDump.cpp]: Sorry, the coding style is different now between Dolphin and PPSSPP, I'll be sure to fix the style issues.

File: Core/HLE/__sceAudio.cpp (hunk 6 above)
Comment (RisingFog) [inline Core/HLE/__sceAudio.cpp]: Made it static.

File: Core/CoreTiming.cpp (hunk 7 above)
Comment (RisingFog) [inline Core/CoreTiming.cpp]: Removed, it was from initial porting which turned out to never needed to be used.

File: Core/AVIDump.cpp (hunk 11 above)
Comment (RisingFog) [inline Core/AVIDump.cpp]: I've exposed it in Screenshot.

File: Core/AVIDump.cpp (hunk 12 above)
Comment (RisingFog) [inline Core/AVIDump.cpp]: Moved.

File: Windows/ppsspp.rc (hunk 15 above)
Comment (RisingFog) [inline Windows/ppsspp.rc]: Fixed.

File: Core/AVIDump.cpp (hunk 12 above)
Comment (RisingFog) [inline Core/AVIDump.cpp]: Made static.

File: Core/AVIDump.cpp (hunk 10 above)
Comment (RisingFog) [inline Core/AVIDump.cpp]: Removed.

File: Core/AVIDump.cpp [hunk 17]
```diff
@@ -0,0 +1,360 @@
...
+#if LIBAVCODEC_VERSION_MAJOR >= 55
+	if (av_frame_get_buffer(s_scaled_frame, 1))
+		return false;
+#else
+	if (avcodec_default_get_buffer(s_stream->codec, s_scaled_frame))
+		return false;
+#endif
+
+	NOTICE_LOG(G3D, "Opening file %s for dumping", s_format_context->filename);
+	if (avio_open(&s_format_context->pb, s_format_context->filename, AVIO_FLAG_WRITE) < 0 || avformat_write_header(s_format_context, nullptr))
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: Does this take filenames in utf-8 or MBCS on Windows?  [According to this it's utf-8](http://stackoverflow.com/questions/6265555/call-avio-open-function-with-non-english-filename-is-invalid) so we should be good (all our file functions live in utf-8 land), but just noting in case.

-[Unknown]

File: Core/HLE/__sceAudio.cpp (hunk 5 above)
Comment (RisingFog) [inline Core/HLE/__sceAudio.cpp]: Made it a static member, and have it clear on audio shutdown.

File: Core/AVIDump.cpp (hunk 13 above)
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: Well, when aligning multiple lines I think it's fine to use spaces for the alignment, just not the indent.  Not the hugest deal though, I think we have some style issues already...

About style, we actually migrated to `if (foo) {` but I think the most important thing is consistency within a file.

-[Unknown]

**Comment by RisingFog** [comment]
@unknownbrackets I don't really agree with renaming Movie to TAS/Record, it's pretty standard these days for emulators to label their TAS tools within a Movie menu of some kind.

File: CMakeLists.txt (hunk 8 above)
Comment (RisingFog) [inline CMakeLists.txt]: It should hopefully be handled now.

File: Core/AVIDump.cpp (hunk 17 above)
Comment (RisingFog) [inline Core/AVIDump.cpp]: I'm pretty sure it's utf-8

File: CMakeLists.txt (hunk 8 above)
Comment (RisingFog) [inline CMakeLists.txt]: Looks like it's not handled, would I have to edit Core.pro itself?

**Comment by unknownbrackets** [comment]
Hmm... I checked a few emulators, none had Movie as top-level menus:

 * higan
 * pcsx2
 * citra
 * rpcs3
 * xebra
 * nestopia
 * snes9x
 * jpcsp
 * no$psx / no$gba
 * vba-m
 * desmume

I saw "Record AVI" under File for some of them, but no top-level Movie.  I'm not disagreeing that people who are into TAS might expect "Movie", but I think that your average user (which there are probably more of) will be confused by "Movie".

None of the emulators I could think of off the top of my head, with the single exception of Dolphin (I didn't check, just presuming) used Movie, so I assume it's not something regular users are used to.  Maybe I'm completely out of touch.

@hrydgard what do you think?

-[Unknown]

**Comment by RisingFog** [comment]
@unknownbrackets Out of those emulators, only Desmume and Dolphin have TAS tools.

**Comment by xnamkcor** [comment]
@unknownbrackets Even among "rr" TAS builds of emulators, input and video "movies" are under File. I can't recall any emulators with Movie or similar names on the menu bar.

**Comment by RisingFog** [comment]
At least with BizHawk, the Movie sub-menu is within File. I could move it into there perhaps.

**Comment by unknownbrackets** [comment]
I think it would mostly make sense under File, sure.  I'm only really against "Movie" if it's at the top level.

-[Unknown]

**Comment by RisingFog** [comment]
@LunaMoo it's going to eventually hold TAS related tools as well, not just for exporting audio/video.

**Comment by LunaMoo** [comment]
As far as I checked it means "tool assisted speedrun" nothing about "movies".

Recording gameplays does not make movies, just like recording theatre play wouldn't, to me it's just wrong word abused in the net because in some languages there might be no distinction between movies and other recordings.

 To me if it would be a main menu in the future filled with "tas" features - it should be simply called "TAS" as [Unknown] proposed, and if it ends up a submenu under File just for video recording, why not call it just that ... "Video recording".

**Comment by unknownbrackets** [comment]
I don't disagree, but whatever naming makes most sense to TAS enthusiasts is probably the best naming, I think.  Maybe the rationale is that TAS is basically creating a "movie" of an ideal play of the game - at least, I think that's the goal of TAS, but I'm not sure.  So in other words, "Trails in the Sky PSP, the Movie".

One way or another, I don't think the name in the menu naming should be an obstacle to merging this functionality.  We can always change it later.  Maybe when there are "Step Frame", "Rewind Step", "Record Gameplay...", "Replay Gameplay...", "Replay Gameplay Until...", "Set Calendar Time...", "Convert Gameplay to AVI...", and whatever other options we will think another name makes sense.

-[Unknown]

**Comment by LunaMoo** [comment]
Agree it's not important, but find it weird that people interested in TAS would not recognize that "TAS" menu is for them;p.

 When I think about PSP "movies" all my eyes see is that page someone requesting UMD-video support linked ~ filled with weird japanese porn.

**Comment by xnamkcor** [comment]
@LunaMoo "Movie" is a term specific to TASs and is used in standard emulators with input/video recording and in TAS specific builds. I get that you want to be pedantic about the words used, but there is an existing nomenclature used pertaining to speedruns. Your unfamiliarity with a use of a word does not devalue its ability to be used in those situations.

**Comment by hrydgard** [comment]
What I don't understand is why display and audio recording are considered TAS-related at all. They can be used perfectly fine even if you're not re-recording, frame stepping and things like that.

Will think a little about where I'd prefer these commands. Note that I'm not necessarily opposed to a new top level menu, but I agree that "movie" is a little strange, and will cause problems for translators.

**Comment by RisingFog** [comment]
I just decided to move it to the File menu and rename it to record.

**Comment by unknownbrackets** [comment]
I'm gonna merge for now, we can always move it later.  I don't think this is a very risky change for the upcoming release.

-[Unknown]

File: Core/AVIDump.cpp [hunk 18]
```diff
@@ -0,0 +1,261 @@
...
+
+	return true;
+}
+
+static void PreparePacket(AVPacket* pkt)
+{
+	av_init_packet(pkt);
+	pkt->data = nullptr;
+	pkt->size = 0;
+	if (s_stream->codec->coded_frame->pts != AV_NOPTS_VALUE)
```
Comment (unknownbrackets) [inline Core/AVIDump.cpp]: It looks like `coded_frame` is deprecated and wasn't reliable.  Could we use `s_stream->pts`?

-[Unknown]
